File: dsa/10+11.graph/101_Symmetric_Tree.py
import typing as t
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSymmetric(self, root: t.Optional[TreeNode]) -> bool:
        return self.isSymmetricHelper(root.left, root.right)
    
    def isSymmetricHelper(self, left: t.Optional[TreeNode], right: t.Optional[TreeNode]) -> bool:
        if not left and not right:
            return True
        
        if not left or not right:
            return False
        
        return (
            left.val == right.val
            and self.isSymmetricHelper(left.left, right.right)
            and self.isSymmetricHelper(left.right, right.left)
        )

File: dsa/10+11.graph/101_Symmetric_Tree.java
package graph;


//  * Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public boolean isSymmetric(TreeNode root) {
        return this.isSymmetricHelper(root.left, root.right);
    }

    public boolean isSymmetricHelper(TreeNode left, TreeNode right){
        if (left == null && right == null){
            return true;
        }

        if (left == null || right == null){
            return false;
        }

        if (left.val != right.val){
            return false;
        }

        return this.isSymmetricHelper(left.left, right.right) && this.isSymmetricHelper(left.right, right.left);
    }
}
